File: companion-apps/ios/CompanionApp/Integration/Geo/GlitchTurnFilter.swift
import Foundation

private let glitchClusterMaxGapM: Double = 10
private let glitchAngleThresholdDeg: Double = 10
private let glitchAngleLookDistanceM: Double = 10

/// Remove clusters of consecutive maneuvers that are map glitches:
/// >=2 maneuvers each within 10m of the previous, where the net path
/// direction change from before the cluster to after it is < 10 degrees.
func filterGlitchClusters(_ maneuvers: [RouteManeuver], geometry: [CoordinatePoint]) -> [RouteManeuver] {
    guard maneuvers.count >= 2, geometry.count >= 2 else { return maneuvers }

    let cumDist = cumulativeDistances(geometry)
    guard !cumDist.isEmpty else { return maneuvers }

    var result = maneuvers
    var i = 0
    while i < result.count {
        guard i + 1 < result.count else { break }
        let gap = result[i + 1].distanceFromStartMeters - result[i].distanceFromStartMeters
        if gap > glitchClusterMaxGapM { i += 1; continue }

        // Extend cluster forward while consecutive gaps <= threshold
        var clusterEnd = i + 1
        while clusterEnd + 1 < result.count {
            let nextGap = result[clusterEnd + 1].distanceFromStartMeters - result[clusterEnd].distanceFromStartMeters
            guard nextGap <= glitchClusterMaxGapM else { break }
            clusterEnd += 1
        }

        let clusterSize = clusterEnd - i + 1
        guard clusterSize >= 2 else { i = clusterEnd + 1; continue }

        // Compute net direction change before -> after the cluster
        if let firstIdx = closestPointIndex(in: geometry, to: result[i].location),
           let lastIdx = closestPointIndex(in: geometry, to: result[clusterEnd].location),
           firstIdx >= 0, firstIdx < geometry.count,
           lastIdx >= 0, lastIdx < geometry.count {

            let entryApproach = walkAlongPolyline(geometry, cumDist, firstIdx, glitchAngleLookDistanceM, direction: "backward")
            let exitDepart = walkAlongPolyline(geometry, cumDist, lastIdx, glitchAngleLookDistanceM, direction: "forward")
            let entryBearing = bearingDegrees(from: entryApproach, to: geometry[firstIdx])
            let exitBearing = bearingDegrees(from: geometry[lastIdx], to: exitDepart)
            var delta = abs(exitBearing - entryBearing)
            if delta > 180 { delta = 360 - delta }

            if delta < glitchAngleThresholdDeg {
                result.removeSubrange(i...clusterEnd)
                continue // re-evaluate this position after removal
            }
        }
        i = clusterEnd + 1
    }
    return result
}

File: companion-apps/web/src/integrations/cues/glitchTurnFilter.ts
import type { CoordinatePoint, RouteManeuver } from "../../domain/models.js";
import {
  bearingDegrees,
  cumulativeDistances,
  findClosestPointIndex,
  walkAlongPolyline,
} from "../geo.js";

const GLITCH_CLUSTER_MAX_GAP_M = 10;
const GLITCH_ANGLE_THRESHOLD_DEG = 10;
const GLITCH_ANGLE_LOOK_DISTANCE_M = 10;

/**
 * Remove clusters of consecutive maneuvers that are map glitches:
 * ≥2 maneuvers each within 10m of the previous, where the net path
 * direction change from before the cluster to after it is < 10°.
 */
export function filterGlitchClusters(
  maneuvers: RouteManeuver[],
  geometry: CoordinatePoint[],
): RouteManeuver[] {
  if (maneuvers.length < 2 || geometry.length < 2) return maneuvers;

  const cumDist = cumulativeDistances(geometry);
  if (cumDist.length === 0) return maneuvers;

  const result = [...maneuvers];
  let i = 0;
  while (i < result.length) {
    if (i + 1 >= result.length) break;
    const gap = result[i + 1].distanceFromStartMeters - result[i].distanceFromStartMeters;
    if (gap > GLITCH_CLUSTER_MAX_GAP_M) {
      i++;
      continue;
    }

    // Extend cluster forward while consecutive gaps ≤ threshold
    let clusterEnd = i + 1;
    while (clusterEnd + 1 < result.length) {
      const nextGap =
        result[clusterEnd + 1].distanceFromStartMeters - result[clusterEnd].distanceFromStartMeters;
      if (nextGap > GLITCH_CLUSTER_MAX_GAP_M) break;
      clusterEnd++;
    }

    const clusterSize = clusterEnd - i + 1;
    if (clusterSize < 2) {
      i = clusterEnd + 1;
      continue;
    }

    // Compute net direction change before→after the cluster
    const firstIdx = findClosestPointIndex(geometry, result[i].location);
    const lastIdx = findClosestPointIndex(geometry, result[clusterEnd].location);

    if (firstIdx >= 0 && firstIdx < geometry.length && lastIdx >= 0 && lastIdx < geometry.length) {
      const entryApproach = walkAlongPolyline(
        geometry,
        cumDist,
        firstIdx,
        GLITCH_ANGLE_LOOK_DISTANCE_M,
        "backward",
      );
      const exitDepart = walkAlongPolyline(
        geometry,
        cumDist,
        lastIdx,
        GLITCH_ANGLE_LOOK_DISTANCE_M,
        "forward",
      );
      const entryBearing = bearingDegrees(entryApproach, geometry[firstIdx]);
      const exitBearing = bearingDegrees(geometry[lastIdx], exitDepart);
      let delta = Math.abs(exitBearing - entryBearing);
      if (delta > 180) delta = 360 - delta;

      if (delta < GLITCH_ANGLE_THRESHOLD_DEG) {
        result.splice(i, clusterSize);
        continue; // re-evaluate this position after removal
      }
    }
    i = clusterEnd + 1;
  }
  return result;
}
